### PR TITLE
Api for notifications

### PR DESCRIPTION
**Pull request by fabrixxm**
and a new class "NotificationsManager"

### CODE REVIEW

**Comment by annando** [comment]
Could you extend the API documentation in doc/api.md?

File: include/NotificationsManager.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
+<?php
+require_once("include/datetime.php");
```
Comment (rabuzarus) [inline include/NotificationsManager.php]:
```
/**
 * @file include/NotificationsManager.php
 */
```

File: include/NotificationsManager.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+class NotificationsManager {
+    private $a;
```
Comment (rabuzarus) [inline include/NotificationsManager.php]:
```
/**
 * @brief Some description of the class
 */

File: include/NotificationsManager.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
+<?php
+require_once("include/datetime.php");
+
```
Comment (rabuzarus) [inline include/NotificationsManager.php]:
```
/**
 * @brief Some description of the class
 */
```

File: include/NotificationsManager.php (hunk 1 above)
Comment (rabuzarus) [inline include/NotificationsManager.php]:
```
/**
 * @file include/NotificationsManager.php
 */
```
above require_once

File: include/NotificationsManager.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    
+    /**
+     * @brief set seen state of $note of local_user()
+     *
+     * @param array $note
+     * @param bool $seen optional true or false, default true
+     * @return bool true on success, false on errors
+     */
+    public function setSeen($note, $seen = true) {
+        return q("update notify set seen = %d where ( link = '%s' or ( parent != 0 and parent = %d and otype = '%s' )) and uid = %d",
```
Comment (rabuzarus) [inline include/NotificationsManager.php]: Should follow the unofficial coding standard for SQL queries 
```
"UPDATE `notify` SET `seen` = %d WHERE ( `link` = '%s' OR ( `parent` != 0 AND `parent` = %d AND `otype` = '%s' )) AND `uid` = %d",
```
maybe some line breaks would be better for readability

File: include/NotificationsManager.php [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    }
+    
+    /**
+     * @brief get one note for local_user() by $id value
+     *
+     * @param int $id
+     * @return array note values or null if not found
+     */
+    public function getByID($id) {
+        $r = q("select * from notify where id = %d and uid = %d limit 1",
```
Comment (rabuzarus) [inline include/NotificationsManager.php]: Should follow the unofficial friendica coding standard for SQL queries
```
"SELECT * FROM `notify` WHERE `id` = %d AND `uid` = %d LIMIT 1",
```

File: include/NotificationsManager.php [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    }
+       
+    /**
+     * @brief set seen state of all notifications of local_user()
+     *
+     * @param bool $seen optional true or false. default true
+     * @return bool true on success, false on error
+     */
+    public function setAllSeen($seen = true) {
+    	return q("update notify set seen = %d where uid = %d",
```
Comment (rabuzarus) [inline include/NotificationsManager.php]: Should follow the unofficial friendica coding standard for SQL queries:
```
"UPDATE `notify` SET `seen` = %d WHERE `uid` = %d"
 ```

File: include/api.php [hunk 7]
```diff
@@ -3386,6 +3414,43 @@ function api_friendica_activity(&$a, $type) {
 	api_register_func('api/friendica/activity/unattendno', 'api_friendica_activity', true, API_METHOD_POST);
 	api_register_func('api/friendica/activity/unattendmaybe', 'api_friendica_activity', true, API_METHOD_POST);
 
+	/**
+	 * returns notifications
+	 * if called with note id set note seen and returns associated item (if possible)
+	 */
```
Comment (rabuzarus) [inline include/api.php]:
```
/**
 * @brief Returns notifications
 *
 * If called with note id set note seen and returns associated item (if possible)
 *
 * @param App $a
 * @param string $type Known types are 'atom', 'rss', 'xml' and 'json'
 */
```
I don't know if we could give a good `@return` description here

File: include/NotificationsManager.php [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+require_once("include/datetime.php");
+
+class NotificationsManager {
+    private $a;
+    
+    public function __construct() {
+        $this->a = get_app();
+    }
+    
+    private function _set_extra($notes) {
```
Comment (rabuzarus) [inline include/NotificationsManager.php]: This should have a short doxygen documentation if possible

**Comment by fabrixxm** [comment]
I've added doxygen comments as @rabuzarus pointed and I've updated docs.
I've separated code for list notifications and for set seen, to set different required http method.
I've not added backticks to queries because is a mysql-only thing to escape problematic names, and I think that if it's not needed it shouldn't be used... but if you think is worth add them, I'll add them.

**Comment by rabuzarus** [comment]
@fabrixxm 
According to the sql queries. It would be really good to have a consistent style here. There are two reasons from my side:
* If you working on a project where you often read much code you get used to this style. Another style make it harder to read. But the more important point is,
* if you edit something somewhere else in the code or at the db scheme (or sometimes only because you want to understand something) you search for code snippet (like with grep). It is very hard to find the needed piece of code by searching if we have too many possibilities how it could look like. We should avoid this where we can. This could lower developing time and some bugs

**Comment by fabrixxm** [comment]
:+1:

File: include/api.php [hunk 9]
```diff
@@ -3386,6 +3419,64 @@ function api_friendica_activity(&$a, $type) {
...
+		$id = (x($_REQUEST, 'id') ? intval($_REQUEST['id']) : 0);
+		
+		$nm = new NotificationsManager();		
+		$note = $nm->getByID($id);
+		if (is_null($note)) throw new BadRequestException("Invalid argument");
+		
+		$nm->setSeen($note);
+		if ($note['otype']=='item') {
+			// would be really better with an ItemsManager and $im->getByID() :-P
+			$r = q("SELECT * FROM item WHERE id=%d AND uid=%d",
```
Comment (rabuzarus) [inline include/api.php]:
```
"SELECT * FROM `item` WHERE `id` = %d AND `uid` = %d",
```

File: mod/notify.php [hunk 10]
```diff
@@ -47,38 +38,37 @@ function notify_init(&$a) {
...
-					'$item_when' => relative_date($it['date'])
-				));
-			}
-		} else {
-			$notif_content .= t('No more system notifications.');
+	$r = $nm->getAll(array('seen'=>0));
+	if ($r!==false && count($r) > 0) {
+		foreach ($r as $it) {
+			$notif_content .= replace_macros($not_tpl,array(
+				'$item_link' => $a->get_baseurl(true).'/notify/view/'. $it['id'],
```
Comment (rabuzarus) [inline mod/notify.php]: just a hint:
we can now use `app::get_baseurl()` instead of `$a->get_baseurl()`  so we don't have to transmit `$a` when the only thing we need from `$a` would be the baseurl

File: include/api.php (hunk 9 above)
Comment (annando) [inline include/api.php]: I, personally, would prefer selecting only the needed fields. This makes queries faster and wastes fewer memory.

File: mod/notify.php (hunk 10 above)
Comment (fabrixxm) [inline mod/notify.php]: that line hasn't been modified, just the indentation. I would leave this kind of fix for another PR.

But anyway, that code is in `mod/notify.php::notify_content(&$a)`, which already get the `$a` value...

File: include/api.php (hunk 9 above)
Comment (fabrixxm) [inline include/api.php]: That line returns an API item object. I could replace that with a query similar to one in `api_statuses_user_timeline()`, on line 1769, but it has "SELECT `item`.* [...]" that too...
Could be a point for a future PR, maybe when introducing an ItemsManager class...

**Comment by annando** [comment]
Could you resolve the conflict? :-)

**Comment by fabrixxm** [comment]
oups ! :-)

**Comment by fabrixxm** [comment]
ready

File: include/NotificationsManager.php [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+	 *  - msg_plain: message as plain text string
+	 */
+    private function _set_extra($notes) {
+        $rets = array();
+        foreach($notes as $n) {
+            $local_time = datetime_convert('UTC',date_default_timezone_get(),$n['date']);
+            $n['timestamp'] = strtotime($local_time);
+            $n['date_rel'] = relative_date($n['date']);
+			$n['msg_html'] = bbcode($n['msg'], false, false, false, false);
+			$n['msg_plain'] = explode("\n",trim(html2plain($n['msg_html'], 0)))[0];
```
Comment (tobiasd) [inline include/NotificationsManager.php]: the html2plain function is unknown at this point on my node which results in a fatal error when following a notification to a posting.

Comment (fabrixxm) [inline include/NotificationsManager.php]: should be fixed by PR #2343

Comment (tobiasd) [inline include/NotificationsManager.php]: it is :-)
